processor/src/advice/dsa.rs

```diff
@@ -0,0 +1,118 @@
+use super::{ExecutionError, Felt, StarkField, Vec, Word};
+use vm_core::crypto::dsa::{
+    elements_as_bytes, Polynomial, PublicKeyBytes, SecretKey, SecretKeyBytes, PK_LEN, SK_LEN,
+};
+
+/// Gets as input a vector containing an expanded public key and its associated secret key, and a
+/// word representing a message and outputs a vector of values to be pushed onto the advice stack.
+/// The values are the ones required for a Falcon signature verification inside the VM and they are:
+///
+/// 1. The nonce represented as 8 field elements.
+/// 2. The expanded public key represented as the coefficients of a polynomial of degree < 512.
+/// 3. The signature represented as the coefficients of a polynomial of degree < 512.
+/// 4. The product of the above two polynomials in the ring of polynomials with coefficients
+/// in the Miden field.
+///
+/// Inputs:
+///   Operand stack: [PK, MSG, ...]
+///   Advice stack: [...]
+///   Advice map: {PK: [pk_raw, sk_raw]}
+///
+/// Outputs:
+///   Operand stack: [PK, MSG, ...]
+///   Advice stack: [NONCE1, NONCE0, h, s2, pi]
+///   Advice map: {PK: [pk_raw, sk_raw]}
+///
+/// Where:
+/// - PK is the digest of an expanded public.
+/// - MSG is the digest of the message to be signed.
+/// - [NONCE0, NONCE1] is a double-word representing a 40 bit nonce that is used in the Falcon
+/// hash-to-point algorithm.
+/// - h is the polynomial representing the expanded public key corresponding to the digest PK.
+/// - s2 is the polynomial representing the signature with the secret key associated to PK on
+/// the message MSG.
+/// - pi is the product of the above two polynomials.
+/// - pk_raw are raw bytes of the expanded public key.
+/// - sk_raw are raw bytes of the secret key.
+///
+/// # Errors
+/// Will return an error if either:
+/// - The advice map does not contain an entry with key PK.
+/// - The advice map entry under key PK is not a vector of the expected length.
+pub fn falcon_sign(pk_sk: &[Felt], msg: Word) -> Result<Vec<Felt>, ExecutionError> {
+    if pk_sk.len() != (PK_LEN + SK_LEN) {
+        return Err(ExecutionError::AdviceStackReadFailed(0));
+    }
```

Comment:
I wonder if we should create a specialized error variant for this. Maybe something like `InvalidKeyPair`?